File: frontend/types/scripts.ts
/** AI Script Intelligence types */

import type { HookPattern } from "@/types/hooks";

export interface ScriptStructure {
  opening_hook: string;
  intro: string;
  key_points: string[];
  transitions: string[];
  cta: string;
  closing: string;
  full_script: string;
}

export interface ScriptAnalytics {
  estimated_engagement: number;
  hook_strength: number;
  emotional_triggers: string[];
  creator_similarity: number;
  readability: number;
  notes: string;
}

export interface ScriptGenerateRequest {
  creator_name: string;
  topic: string;
  tone: string;
  duration: string;
  hook_type: string;
}

export interface ScriptGenerateResult {
  creator_name: string;
  topic: string;
  tone: string;
  duration: string;
  hook_type: string;
  selected_hook: string;
  structure: ScriptStructure;
  analytics: ScriptAnalytics;
  viral_hooks_used: string[];
  style_notes: string;
}

export interface ScriptAnalyzeRequest {
  creator_name?: string;
  script_text?: string;
  topic?: string;
}

export interface ScriptAnalyzeResult {
  creator_name: string;
  summary: string;
  structure_detected: ScriptStructure;
  analytics: ScriptAnalytics;
  recommendations: string[];
}

export interface ScriptCompareRequest {
  creator_name: string;
  generated_script: string;
  topic?: string;
}

export interface ScriptCompareResult {
  summary: string;
  style_alignment: string;
  hook_alignment: string;
  gaps: string[];
  strengths: string[];
  top_video_references: string[];
}

export interface CreatorStyleContext {
  creator_name: string;
  content_style: string;
  communication_style: string;
  top_topics: string[];
  hook_patterns: string[];
  vocabulary: string[];
  sample_titles: string[];
  transcript_excerpts: string[];
}

export interface ScriptWorkspace {
  creators: string[];
  default_structure: ScriptStructure;
  viral_hooks: HookPattern[];
  creator_style: CreatorStyleContext | null;
  structure_template_notes: string;
}

export interface GeneratedScriptEntry {
  id: string;
  result: ScriptGenerateResult;
  createdAt: string;
}

File: backend/app/schemas/scripts.py
"""Pydantic schemas for AI Script Intelligence."""

from pydantic import BaseModel, Field

from app.schemas.hooks import HookPatternRead


class ScriptStructure(BaseModel):
    """Full script outline — opening through closing."""

    opening_hook: str = ""
    intro: str = ""
    key_points: list[str] = Field(default_factory=list)
    transitions: list[str] = Field(default_factory=list)
    cta: str = ""
    closing: str = ""
    full_script: str = ""


class ScriptAnalytics(BaseModel):
    """
    Explainable quality scores (0–100) for a generated script.

    Computed from hook match, creator vocabulary overlap, and readability heuristics.
    """

    estimated_engagement: float = 0.0
    hook_strength: float = 0.0
    emotional_triggers: list[str] = Field(default_factory=list)
    creator_similarity: float = 0.0
    readability: float = 0.0
    notes: str = ""


class ScriptGenerateRequest(BaseModel):
    """User inputs for creator-aware script generation."""

    creator_name: str = Field(..., min_length=1)
    topic: str = Field(..., min_length=2)
    tone: str = "conversational"
    duration: str = "10 minutes"
    hook_type: str = "curiosity"


class ScriptGenerateResult(BaseModel):
    """Complete generated script + analytics + hooks used."""

    creator_name: str
    topic: str
    tone: str
    duration: str
    hook_type: str
    selected_hook: str = ""
    structure: ScriptStructure = Field(default_factory=ScriptStructure)
    analytics: ScriptAnalytics = Field(default_factory=ScriptAnalytics)
    viral_hooks_used: list[str] = Field(default_factory=list)
    style_notes: str = ""


class ScriptAnalyzeRequest(BaseModel):
    """Analyze script text or creator transcript style."""

    creator_name: str = ""
    script_text: str = ""
    topic: str = ""


class ScriptAnalyzeResult(BaseModel):
    """Analysis of a script or creator speaking style."""

    creator_name: str = ""
    summary: str = ""
    structure_detected: ScriptStructure = Field(default_factory=ScriptStructure)
    analytics: ScriptAnalytics = Field(default_factory=ScriptAnalytics)
    recommendations: list[str] = Field(default_factory=list)


class ScriptCompareRequest(BaseModel):
    """Compare generated script vs creator baseline."""

    creator_name: str
    generated_script: str
    topic: str = ""


class ScriptCompareResult(BaseModel):
    """Generated vs creator style / top videos."""

    summary: str = ""
    style_alignment: str = ""
    hook_alignment: str = ""
    gaps: list[str] = Field(default_factory=list)
    strengths: list[str] = Field(default_factory=list)
    top_video_references: list[str] = Field(default_factory=list)


class CreatorStyleContext(BaseModel):
    """Creator DNA for /scripts workspace — hooks, patterns, vocabulary."""

    creator_name: str
    content_style: str = ""
    communication_style: str = ""
    top_topics: list[str] = Field(default_factory=list)
    hook_patterns: list[str] = Field(default_factory=list)
    vocabulary: list[str] = Field(default_factory=list)
    sample_titles: list[str] = Field(default_factory=list)
    transcript_excerpts: list[str] = Field(default_factory=list)


class ScriptWorkspace(BaseModel):
    """GET /scripts/workspace — page context before generation."""

    creators: list[str] = Field(default_factory=list)
    default_structure: ScriptStructure = Field(default_factory=ScriptStructure)
    viral_hooks: list[HookPatternRead] = Field(default_factory=list)
    creator_style: CreatorStyleContext | None = None
    structure_template_notes: str = ""


class ScriptGenerationIntel(BaseModel):
    """LangGraph structured output for script_generation."""

    topic: str = ""
    creator_name: str = ""
    selected_hook: str = ""
    structure: ScriptStructure = Field(default_factory=ScriptStructure)
    analytics: ScriptAnalytics = Field(default_factory=ScriptAnalytics)
    style_notes: str = ""
    recommendations: list[str] = Field(default_factory=list)


class ScriptAnalysisIntel(BaseModel):
    """LangGraph structured output for script_analysis."""

    summary: str = ""
    analytics: ScriptAnalytics = Field(default_factory=ScriptAnalytics)
    recommendations: list[str] = Field(default_factory=list)
    structure_notes: str = ""
